test/tests.py

```diff
@@ -1508,6 +1509,154 @@ def test(self):
+            if dml_lineno is not None:
+                dml_lineno += 1
+            line = line.rstrip()
+            m = line_directive_re.match(line)
+            if m is not None:
+                (lineno, filename) = m.groups()
+                lineno = int(lineno)
+                if (Path(self.scratchdir) / filename == c_file):
+                    assert not reset, \
+                        f"Redundant line directive reset at {c_lineno}"
+                    assert lineno == c_lineno + 1, \
+                        f"Incorrect line directive reset at {c_lineno}"
+                    reset = True
+                    dml_lineno = None
+                elif filename.endswith(self.filename):
+                    assert dml_lineno != lineno, \
+                            (f'Redundant linemark for {dml_lineno} at '
+                             + f'{c_lineno}')
+                    reset = False
+                    dml_lineno = lineno - 1
+                else:
+                    reset = False
+                    dml_lineno = None
+                continue
+            m = func_start_re.match(line)
+            if m is not None:
+                assert curr_func is None
+                curr_func = m.group(1)
+                assert reset, \
+                    f'Line directive not reset by start of {curr_func}'
+                continue
+            if dml_lineno is not None:
+                yield (dml_lineno, c_lineno, curr_func)
+            if curr_func is not None and line == '}':
+                curr_func = None
+
+        assert reset, "Line directive not reset by end of C file"
+        assert curr_func is None, "Unterminated C function"
+
+
+    def test(self):
+        super().test()
+        c_file = Path(f'{self.cfilename}.c').resolve()
+        h_file = Path(f'{self.cfilename}.h').resolve()
+        base_file = Path(self.filename).resolve()
+
+        func_spans = {}
+        cblock_spans = {}
+        not_linemarked = set()
+        with open(base_file) as base_file_handle:
+            curr_span = None
+            for (lineno, line) in enumerate(base_file_handle, 1):
+                line = line.strip()
+                m = begin_linemarks_func_re.match(line)
+                if m:
+                    assert curr_span is None, f"{lineno}: {curr_span}"
+                    curr_span = (func_spans, m.group(1).split(), lineno + 1)
+                elif line == '// BEGIN-LINEMARKS-CBLOCK':
+                    curr_span = (cblock_spans, lineno + 1)
+                elif line == '// END-LINEMARKS':
+                    assert curr_span is not None
+                    if curr_span[0] is func_spans:
+                        for func in curr_span[1]:
+                            func_spans.setdefault(func, set()).add(
+                                (curr_span[2], lineno - 1))
+                    else:
+                        cblock_spans[curr_span[1]] = lineno - 1
```

Comment:
Consider extracting this DML analysis to a function (and maybe the same for the C code analysis), to reduces the number of mutated variables simultaneously in scope.

Reply:
Good idea.

Reply:
Done both.